pkg/gardener/shoot/extender/provider/provider.go

```diff
@@ -78,20 +74,40 @@ func NewProviderExtenderPatchOperation(enableIMDSv2 bool, defMachineImgName, def
 			provider.Workers = append(provider.Workers, *rt.Spec.Shoot.Provider.AdditionalWorkers...)
 		}
 
-		controlPlaneConf, infraConfig := overrideConfigIfProvided(rt, existingInfraConfig, existingControlPlaneConfig)
+		provider.Workers = sortWorkersToShootOrder(provider.Workers, shootWorkers)
 
-		workerZones, err := getNetworkingZonesFromWorkers(provider.Workers)
+		workerZonesFromRuntime, err := getNetworkingZonesFromWorkers(provider.Workers)
 		if err != nil {
 			return err
 		}
 
-		// final validation
-		if err = checkWorkerZonesMatchProviderConfig(rt.Spec.Shoot.Provider.Type, workerZones, controlPlaneConf, infraConfig, true); err != nil {
+		workerZonesFromShoot, err := getNetworkingZonesFromWorkers(shootWorkers)
+		if err != nil {
 			return err
 		}
 
-		provider.ControlPlaneConfig = controlPlaneConf
-		provider.InfrastructureConfig = infraConfig
+		zonesAdded := newZonesAdded(workerZonesFromShoot, workerZonesFromRuntime)
+
+		if zonesEqual(workerZonesFromShoot, workerZonesFromRuntime) || len(zonesAdded) == 0 {
+			provider.ControlPlaneConfig = existingControlPlaneConfig
+			provider.InfrastructureConfig = existingInfraConfig
+		} else {
+			mergedWorkerZones := append(workerZonesFromShoot, zonesAdded...)
+
+			infraConfig, controlPlaneConfig, err := getConfig(rt.Spec.Shoot, mergedWorkerZones, existingInfraConfig.Raw)
```

Comment:
Idea to consider- do not modify generation of config and use data from existing config to enchance - just copy all existing data without zones configuration: 

``` golang
			mergedWorkerZones := mergeNewZones(workerZonesFromShoot, workerZonesFromRuntime)
			newInfraConfig, newControlPlaneConfig, err := getConfig(rt.Spec.Shoot, mergedWorkerZones)
			infraConfig, controlPlaneConfig = patchConfigWithDataFromShoot(newInfraConfig, newControlPlaneConfig, existingInfraConfig, existingControlPlaneConfig)


```

Reply:
As agreed I will skip this one.